src/valid/function.rs

```diff
@@ -325,24 +333,37 @@ impl super::Validator {
                     // special handling for block scoping is needed here,
                     // because the continuing{} block inherits the scope
                     let base_expression_count = self.valid_expression_list.len();
+                    let return_ability = context.abilities & ControlFlowAbility::RETURN;
                     stages &= self.validate_block_impl(
                         body,
-                        &context.with_flags(Flags::CAN_JUMP | Flags::IN_LOOP),
+                        &context.with_abilities(
+                            return_ability
+                                | ControlFlowAbility::BREAK
+                                | ControlFlowAbility::CONTINUE,
+                        ),
+                    )?;
+                    stages &= self.validate_block_impl(
+                        continuing,
+                        &context.with_abilities(ControlFlowAbility::empty()),
                     )?;
-                    stages &=
-                        self.validate_block_impl(continuing, &context.with_flags(Flags::empty()))?;
                     for handle in self.valid_expression_list.drain(base_expression_count..) {
                         self.valid_expression_set.remove(handle.index());
                     }
                 }
-                S::Break | S::Continue => {
-                    if !context.flags.contains(Flags::IN_LOOP) {
-                        return Err(FunctionError::BreakContinueOutsideOfLoop);
+                S::Break => {
+                    if !context.abilities.contains(ControlFlowAbility::BREAK) {
+                        return Err(FunctionError::BreakOutsideOfLoopOrSwitch);
+                    }
+                    finished = true;
+                }
+                S::Continue => {
+                    if !context.abilities.contains(ControlFlowAbility::CONTINUE) {
+                        return Err(FunctionError::ContinueOutsideOfLoop);
                     }
                     finished = true;
                 }
                 S::Return { value } => {
-                    if !context.flags.contains(Flags::CAN_JUMP) {
+                    if !context.abilities.contains(ControlFlowAbility::RETURN) {
                         return Err(FunctionError::InvalidReturnSpot);
                     }
```

Comment:
Are `Kill` statements permitted in `Loop` `continuing` blocks?

Reply:
I assume it is